### DISCUSSION TITLE
How do I start a new session as not nested

### DISCUSSION THREAD

**Discussion by Wezea** [zellij-org/zellij#2448] [Q&A] [answered]
Hello folks

what a great tool you have created here. Thank you for that great piece of software.

Maybe I've missed it in the docs or I don't understand the way how it is supposed to work.
If I want to start a new session out of an open session `zellij -s foobar`, it will always get nested in the parent.

How do I create a new sibling beside the existing one?

Cheers!

**Comment by Lockszmith-GH** [comment]
Are you asking "how to create a new pane?"

I don't fully know where you're coming from (OS, shell, terminal emulator and your personal experience), so I'll try to answer 'the long way' assuming nothing - and maintaining a vocabulary. I hope it will be helpful, but apologize if not.

The way I understand it, a **session** is "everything on the current terminal connection".

My terminal emulator is either Alacritty on Pop!_OS or Windows Terminal on Windows 11.  
On Linux I will open the terminal which would start `bash` that by default starts my session of Zellij.  
On Windows I will open the terminal and `ssh` into a Linux machine (or WSL2) which starts `bash` (and Zellij).

As an example, let's take the Linux route, after which I'll be in a Zellij **session**, which has opened a **layout** containing among other a **pane** that runs `bash` within it.

If I SSH into the machine (let's say from Windows), depending on how I configured this, I can either open a new **session**, or connect to that same **session**.

A new **session** will have its own set **layouts** and **panes**, which the same session will show the same content on two connections.

From within a session, I might want to open a new **pane** (`Ctrl`+`p` and then `n`) or **tab** (`Ctrl`+`t` and then `n`).

**Comment by Wezea** [comment]
Sorry, I should have been more clear what I want to achieve.

I have multiple work environments and I do want to manage them using separate sessions. Similar like the [tmux-sessionizer](https://github.com/jrmoulton/tmux-sessionizer) achieving this under tmux. The final goal would be a floating pane, where I can easily create new sessions or switch between existing ones.

![Screenshot from 2023-05-15 22-52-28](https://github.com/zellij-org/zellij/assets/5039915/b3964908-98a9-42ac-bccc-557def3336f3)

For work I still use my tmux setup, but I am currently experimenting using a zellij playground. I do also start up zellij when my bash starts. However, once I am in my zellij session, I can't figure out how to start up a new session beside the current, it will always get nested.  I guess [this PR](https://github.com/zellij-org/zellij/pull/2049) point to the direction I want.

**Reply by Lockszmith-GH** [reply]
ooh I see, interesting - yeah, other than detaching and reattaching I don't see an option like that. If/When you figure this out, let us know here. This sounds intersting.

Like you before Zellij I used tmux, but I did not know about tmux-sessionizer.

**Reply by mahume** [reply]
I've got this same workflow. Kind of the only thing holding me back from switching from Tmux to Zellij at this point. For Tmux, I've got a script that opens multiple sessions with different layouts. Tinkering around with Zellij and it seems that the sessions aren't quite there yet. Not really seeing anything in the docs about sessions either. Running into issues with nested sessions and sessions that are disjointed. Have to run `zellij ka` twice. Another issue I noticed was that when opening a session with a custom layout, each new session created uses that same layout, which isn't always desirable when setting up a default layout for all your servers with work.

**Reply by Wezea** [reply]
It is almost [done](https://github.com/zellij-org/zellij/pull/2801), a solution is about to come.

**Comment by robrecord** [comment]
zellij-runner is a tool that you could ask your terminal to invoke whenever it creates a new window or tab. Then you get to choose your session, or create if it doesn't exist. That's what I'm doing and it works well.
https://lib.rs/crates/zellij-runner

**Reply by Lockszmith-GH** [reply]
Interesting, I see your dotfiles are for managing a nix system. Are you using NixOS? or just the Nix package manager?
Any good resources for begining - I tried fiddling with Nix and somehow always get stuck

**Reply by Wezea** [reply]
Great, thank you, I will play with this tool a bit. I probably need to start the runner separately, as if I start it from a running session, and create new session from that point, I would again instantiate nested sessions, wouldn't I?

**Reply by robrecord** [reply]
Yes - You would have to have your terminal start a shell with just this command, rather than a zellij session. So the zellij session would be nested inside the zellij-runner command.

**Reply by robrecord** [reply]
@Lockszmith-GH I don’t think those are my dotfiles you’re seeing, I’m on macOS. Wish I could help.

**Reply by Lockszmith-GH** [reply]
Oh, my bad - I thought you were the maintainer of zellij-runner, so sorry about that.

**Comment by di-rs** [comment]
Hey!

You can check this zellij sessionizer solution [demestoss/zellij-smart-sessionizer](https://github.com/demestoss/zellij-smart-sessionizer). I have started working on it after I faced the same issue as you

It doesn't have functionality for switching sessions inside Zellij itself and currently works with Zoxide setup only, but it could be pretty handy in your use case

You can start sessions from the shell the same way and choose the session using fzf. After that, you can choose the preferred layout for the session

If you run it inside Zellij, it will create a new tab instead of a new session

I'm working on improving it now and I plan to rewrite it and integrate it as a Zellij plugin, so I want it to have the same basic functionality as tmux-sessionizer and even more

**Reply by kamilkrz** [reply]
Hi, nice one!
As i am also waiting for zellij to have detach cmd I was tinkering with same idea. 
Do you consider adding -d'/' --with-nth=-1 to your layout picker?
It seems that you are querying default path anyway so leaving only filename should look better?

**Comment by imsnif** [comment, marked as answer]
Zellij now has this functionality out of the box. There is no need to hack it yourself through the CLI API or anything like that.

Check out the `session-manager` built-in plugin. You can launch it with `Ctrl o` + `w`.

**Reply by kamilkrz** [reply]
As You can see in https://github.com/zellij-org/zellij/discussions/2448#discussioncomment-5909662
there is more to that. 

The goal is to have idempotent way of managing sessions directly tied to custom set of pwd's. For this we got already `attach --create`

This was also mentioned in #365 (https://github.com/zellij-org/zellij/discussions/365#discussioncomment-7544329)

I am also aware that #2962 exists  but I am afraid that simple detaching and attaching again will not work as rest of the script will be bound to zellij session and will produce nested sessions.

Right now I got those to running but still need last part (or parts depends how and what options will be introduced)


```
#selected_name is name of session based on chosen pwd
 
#create or attach no matter what outside of zellij
if [[ -z $ZELLIJ ]]; then
	zellij a --create $selected_name options --default-cwd $selected
	exit 0
fi

#ignore if already in the same session
if [[ $ZELLIJ_SESSION_NAME == $selected_name ]]; then
	exit 0
fi

# below is part waiting for session switch
#???
```


HTH

**Reply by imsnif** [reply]
So, I don't think this will ever be a thing through the CLI, I'm afraid. Zellij has true multiplayer support - meaning several users can be connected to the same session, each having their own cursor. For this reason, actions that affect only a single user cannot happen (well) through the CLI. Switching sessions is a very good example of this. Zellij cannot know to switch one user rather than the other, because the CLI is not aware of multiple users.

You should be able to do this through the plugin API. Plugins are aware of multiple users and for that reason the relevant APIs exist. I realize many users prefer writing scripts for these sorts of things, but I'm afraid this is just not the Zellij way. We encourage users to write plugins and try to make the process as easy as possible. For sure, there's still a way to go, but this is my main focus these days.

P.S.
If someone is wondering how this works in tmux, the answer is "not well". Try to do some experiments with connecting with multiple users to a single tmux session, sleeping && detaching for one user and then doing something with one of the other users before the detach happens. Often, the right thing will not happen. In tmux this is not so bad, because it doesn't have true multiplayer and the chances of this causing harm to someone's workflow are low, this is not the case with Zellij.

**Reply by kamilkrz** [reply]
Oh,
I believe this answer should be marked and not @Wezea then.
Good news, its doable.
This answer should be presented to all questions in discussion and issues that mentions ThePrimeagen's (and derivative) work.
Thanks!

**Reply by zMenta** [reply]
Is there a way that I could call the action the session-manager uses to switch sessions? As "zellij call-plugin session-manager action switch-session $session_name" type of deal?

Because the session-manager has this functionality, but how I can invoke this action trough the plugin API? Is this possible?

**Reply by imsnif** [reply]
@zMenta check out [plugin pipes](https://zellij.dev/documentation/plugin-pipes) for calling/invoking plugins from the CLI (not currently implemented for the session-manager, can totally be implemented for your plugin).

And the [switch session](https://zellij.dev/documentation/plugin-api-commands#switch_session) API call.

This should get you sorted.

**Reply by zMenta** [reply]
Thanks a lot for the reply @imsnif. It's a bummer on my point of view that is not that easy to just call "switch session" from the CLI, also that the pipe is not implemented for the session-manager. 

Anyway, thanks for the reply, I'll check it out in more details when I have more time. But in the current moment I'm happy with my "zellij session script". Otherwise I'll try to just use the session-manager plugin when I'm already inside a zellij session for the occasion when I want to switch sessions.

Good luck on the project! :)
